spring-data-mongodb/src/main/java/org/springframework/data/mongodb/observability/DefaultMongoHandlerObservationConvention.java

```diff
@@ -15,34 +15,75 @@
  */
 package org.springframework.data.mongodb.observability;
 
-import io.micrometer.common.KeyValue;
-import io.micrometer.common.KeyValues;
+import java.net.InetSocketAddress;
 
 import org.springframework.data.mongodb.observability.MongoObservation.HighCardinalityCommandKeyNames;
 import org.springframework.data.mongodb.observability.MongoObservation.LowCardinalityCommandKeyNames;
+import org.springframework.lang.Nullable;
+import org.springframework.util.ObjectUtils;
 
+import com.mongodb.ConnectionString;
+import com.mongodb.ServerAddress;
 import com.mongodb.connection.ConnectionDescription;
 import com.mongodb.connection.ConnectionId;
 import com.mongodb.event.CommandStartedEvent;
 
+import io.micrometer.common.KeyValue;
+import io.micrometer.common.KeyValues;
+
 /**
  * Default {@link MongoHandlerObservationConvention} implementation.
  *
  * @author Greg Turnquist
- * @since 4.0.0
+ * @since 4
  */
-public class DefaultMongoHandlerObservationConvention implements MongoHandlerObservationConvention {
+class DefaultMongoHandlerObservationConvention implements MongoHandlerObservationConvention {
 
 	@Override
 	public KeyValues getLowCardinalityKeyValues(MongoHandlerContext context) {
 
-		KeyValues keyValues = KeyValues.empty();
+		KeyValues keyValues = KeyValues.of(LowCardinalityCommandKeyNames.DB_SYSTEM.withValue("mongodb"));
+
+		ConnectionString connectionString = context.getConnectionString();
+		if (connectionString != null) {
 
-		if (context.getCollectionName() != null) {
+			keyValues = keyValues
+					.and(LowCardinalityCommandKeyNames.DB_CONNECTION_STRING.withValue(connectionString.getConnectionString()));
+
+			String user = connectionString.getUsername();
+
+			if (!ObjectUtils.isEmpty(user)) {
+				keyValues = keyValues.and(LowCardinalityCommandKeyNames.DB_USER.withValue(user));
+			}
+
+		}
+
+		if (!ObjectUtils.isEmpty(context.getDatabaseName())) {
+			keyValues = keyValues.and(LowCardinalityCommandKeyNames.DB_NAME.withValue(context.getDatabaseName()));
+		}
+
+		if (!ObjectUtils.isEmpty(context.getCollectionName())) {
 			keyValues = keyValues
 					.and(LowCardinalityCommandKeyNames.MONGODB_COLLECTION.withValue(context.getCollectionName()));
 		}
 
+		ServerAddress serverAddress = context.getCommandStartedEvent().getConnectionDescription().getServerAddress();
+
+		if (serverAddress != null) {
+
+			keyValues = keyValues.and(LowCardinalityCommandKeyNames.NET_TRANSPORT.withValue("IP.TCP"),
+					LowCardinalityCommandKeyNames.NET_PEER_ADDR.withValue(serverAddress.getHost()));
+
+			InetSocketAddress socketAddress = serverAddress.getSocketAddress();
```

Comment:
I don't remember but wasn't one of this calls very slow?

Reply:
Socket addresses must be resolved at this time already, so we use a cached instance.